compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -605,13 +606,31 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
       val app1 =
         if (!success) app0.withType(UnspecifiedErrorType)
         else {
-          if (!sameSeq(args, orderedArgs) && !isJavaAnnotConstr(methRef.symbol)) {
+          if (!sameSeq(args, orderedArgs.dropWhile(_ eq EmptyTree)) && !isJavaAnnotConstr(methRef.symbol)) {
             // need to lift arguments to maintain evaluation order in the
             // presence of argument reorderings.
+
             liftFun()
-            val eqSuffixLength = firstDiff(app.args.reverse, orderedArgs.reverse)
-            val (liftable, rest) = typedArgs splitAt (typedArgs.length - eqSuffixLength)
-            typedArgs = liftArgs(liftedDefs, methType, liftable) ++ rest
+
+            // lift arguments in the definition order
+            val argDefBuff = mutable.ListBuffer.empty[Tree]
+            typedArgs = liftArgs(argDefBuff, methType, typedArgs)
+
+            def orderedArgDefs = {
+              val impureArgs = typedArgBuf.filterNot { // pure are not lifted by liftArgs
+                case NamedArg(_, arg) => isPureExpr(arg)
```

Comment:
Let's instead add a case for `NamedArg` in `TreeInfo.exprPurity`. That will avoid the special treatment here.